draft-ietf-tls-esni.md

```diff
@@ -1725,7 +1751,8 @@ application layer.
 # Linear-time Outer Extension Processing {#linear-outer-extensions}
 
 The following procedure processes the "ech_outer_extensions" extension (see
-{{encoding-inner}}) in linear time:
+{{encoding-inner}}) in linear time, ensuring that each referenced extension
+in the ClientHelloOuter is included at most once:
```

Comment:
I claim that the algorithm below doesn't actually enforce this. It's straightforward to do so, however. How about just adding a bullet just below the second step that says "If E already appeared previously in OuterExtensions, then abort with "illegal_parameter"."

Reply:
Or, based on the security considerations above, perhaps the intent is just to skip duplicate extensions?

Reply:
This language is intended to be compatible with or without duplicate extensions in the client hello outer. 

If the reader is assuming extension ids are unique, and rejects duplicate extensions, the linear scan can't include the same extension twice and will bail. 

If the reader is assuming duplicate extensions are tolerated, they must treat `ExtFoo ExtFoo` as two different extensions, but the same requirement still works, as in this reading the same extension isn't included twice and there's no amplification attack and no panic. 

If the TLS1.3 RFC is ever update to clarify this, the language here can be tightened to point to that, without impacting implementations :-).

Reply:
Ok.

Reply:
+1 to @dennisjackson's interpretation. Adding something like "If E already appeared previously in OuterExtensions" requires auxiliary state and is redundant with whatever the ClientHelloOuter story is.